ddl/ddl_api.go

```diff
@@ -6298,6 +6300,126 @@ func (d *ddl) AlterTablePartitionAttributes(ctx sessionctx.Context, ident ast.Id
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Voters = op.UintValue
+			case ast.TableOptionPlacementLearnerCount:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Learners = op.UintValue
+			case ast.TableOptionPlacementSchedule:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Schedule = op.StrValue
+			case ast.TableOptionPlacementConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Constraints = op.StrValue
+			case ast.TableOptionPlacementLeaderConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.LeaderConstraints = op.StrValue
+			case ast.TableOptionPlacementLearnerConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.LearnerConstraints = op.StrValue
+			case ast.TableOptionPlacementFollowerConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.FollowerConstraints = op.StrValue
+			case ast.TableOptionPlacementVoterConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.VoterConstraints = op.StrValue
+			default:
+				return errors.Trace(errors.New("unknown placement policy option"))
+			}
+		}
+	}
+
+	// Can not use both a placement policy and direct assignment. If you alter specify both in a CREATE TABLE or ALTER TABLE an error will be returned.
+	if ptDef.DirectPlacementOpts != nil && ptDef.PlacementPolicyRef != nil {
+		return errors.Trace(ErrPlacementPolicyWithDirectOption.GenWithStackByArgs(ptDef.PlacementPolicyRef.Name))
+	}
+	if ptDef.DirectPlacementOpts != nil {
+		// check the direct placement option compatibility.
+		if err := checkPolicyValidation(ptDef.DirectPlacementOpts); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	if ptDef.PlacementPolicyRef != nil {
+		// placement policy reference will override the direct placement options.
+		policy, ok := ctx.GetInfoSchema().(infoschema.InfoSchema).PolicyByName(ptDef.PlacementPolicyRef.Name)
+		if !ok {
+			return errors.Trace(infoschema.ErrPlacementPolicyNotExists.GenWithStackByArgs(ptDef.PlacementPolicyRef.Name))
+		}
+		ptDef.PlacementPolicyRef.ID = policy.ID
+	}
+	job := &model.Job{
+		SchemaID:   schema.ID,
+		TableID:    meta.ID,
+		SchemaName: schema.Name.L,
+		Type:       model.ActionAlterTablePartitionPolicy,
+		BinlogInfo: &model.HistoryInfo{},
+		Args:       []interface{}{partitionID, ptDef},
```

Comment:
I suggest passing `PlacementPolicyRef, DirectPlacementOpts` directly, instead of `ptDef`.

Reply:
donr